### PR TITLE
✨informs customer with role x if company user was created

### PR DESCRIPTION
**Pull request by julianhyatt**
- adds new post create plugin (company user creation)
- checks for configured role
- send notification mail to customer with configured role

### CODE REVIEW

File: bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Business/Model/Mail/CompanyUserCreationNotificationMailHandler.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+     */
+    public function sendInformationMail(CompanyUserTransfer $companyUserTransfer): CompanyUserTransfer
+    {
+        $customerTransfer = $companyUserTransfer->getCustomer();
+
+        if ($customerTransfer->getIsNew() !== true || $this->checkRoles($companyUserTransfer->getCompanyRoleCollection()) !== true) {
+            return $companyUserTransfer;
+        }
+
+        $inform = $this->repository->getNotificationCustomerByFkCompanyAndRole($companyUserTransfer->getFkCompany(), $this->config->getRolesToNotify());
```
Comment (DerCalli) [inline bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Business/Model/Mail/CompanyUserCreationNotificationMailHandler.php:58]: The line seems a bit long to me.

File: bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Persistence/CompanyUserMailConnectorRepository.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+use Spryker\Zed\Kernel\Persistence\AbstractRepository;
+
+/**
+ * @method \FondOfOryx\Zed\CompanyUserMailConnector\Persistence\CompanyUserMailConnectorPersistenceFactory getFactory()
+ */
+class CompanyUserMailConnectorRepository extends AbstractRepository implements CompanyUserMailConnectorRepositoryInterface
+{
+    /**
+     * @param int $fkCompany
+     * @param array $roleNames
```
Comment (DerCalli) [inline bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Persistence/CompanyUserMailConnectorRepository.php]: `@param array<string> $roleNames?`

**Review by DerCalli — APPROVED** [review]

File: bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Business/Model/Mail/CompanyUserCreationNotificationMailHandler.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+        foreach ($inform->getNotificationCustomers() as $notificationCustomer) {
+            $mailTransfer = new MailTransfer();
+            $mailTransfer->setType(CompanyUserWasCreatedInformerMailTypePlugin::MAIL_TYPE);
+            $mailTransfer->setCustomer($customerTransfer);
+            $mailTransfer->setLocale($customerTransfer->getLocale());
+            $mailTransfer->setNotifyCustomer($notificationCustomer);
```
Comment (daniel-rose) [inline bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Business/Model/Mail/CompanyUserCreationNotificationMailHandler.php]:
```suggestion
            $mailTransfer = (new MailTransfer())
				->setType(CompanyUserWasCreatedInformerMailTypePlugin::MAIL_TYPE)
            	->setCustomer($customerTransfer)
            	->setLocale($customerTransfer->getLocale())
            	->setNotifyCustomer($notificationCustomer);
```

File: bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Business/Model/Mail/CompanyUserCreationNotificationMailHandlerInterface.php [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+use Generated\Shared\Transfer\CompanyUserTransfer;
+
+interface CompanyUserCreationNotificationMailHandlerInterface
+{
+    /**
+     * @param \Generated\Shared\Transfer\CompanyUserTransfer $companyUserTransfer
+     *
+     * @return \Generated\Shared\Transfer\CompanyUserTransfer
+     */
+    public function sendInformationMail(CompanyUserTransfer $companyUserTransfer): CompanyUserTransfer;
```
Comment (daniel-rose) [inline bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Business/Model/Mail/CompanyUserCreationNotificationMailHandlerInterface.php]: What is an information mail?

File: bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Persistence/CompanyUserMailConnectorRepository.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        $customerQuery
+            ->useCompanyUserQuery()
+            ->filterByFkCompany($fkCompany)
+            ->useSpyCompanyRoleToCompanyUserQuery()
+            ->useCompanyRoleQuery()
+            ->filterByName_In($roleNames)
+            ->endUse()
+            ->endUse()
+            ->endUse()
+            ->select([SpyCustomerTableMap::COL_FIRST_NAME, SpyCustomerTableMap::COL_LAST_NAME, SpyCustomerTableMap::COL_EMAIL, SpyCompanyRoleTableMap::COL_NAME]);
```
Comment (daniel-rose) [inline bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Persistence/CompanyUserMailConnectorRepository.php]: Use a better code style

File: bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Persistence/CompanyUserMailConnectorRepository.php [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+        return $collection;
+    }
+
+    /**
+     * @param string $field
+     *
+     * @return string
+     */
+    protected function cleanFieldName(string $field): string
```
Comment (daniel-rose) [inline bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Persistence/CompanyUserMailConnectorRepository.php]: Where is this method use?

File: bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Persistence/CompanyUserMailConnectorRepository.php [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+        return $this->createCompanyUserCollection($customerQuery->find()->getData());
+    }
+
+    /**
+     * @param array $data
+     *
+     * @return \Generated\Shared\Transfer\NotificationCustomerCollectionTransfer
+     */
+    protected function createCompanyUserCollection(array $data): NotificationCustomerCollectionTransfer
```
Comment (daniel-rose) [inline bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Persistence/CompanyUserMailConnectorRepository.php]: By what?

File: bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Business/CompanyUserMailConnectorFacadeInterface.php [hunk 8]
```diff
@@ -12,4 +12,11 @@ interface CompanyUserMailConnectorFacadeInterface
      * @return \Generated\Shared\Transfer\CompanyUserTransfer
      */
     public function sendMail(CompanyUserTransfer $companyUserTransfer): CompanyUserTransfer;
+
+    /**
+     * @param \Generated\Shared\Transfer\CompanyUserTransfer $companyUserTransfer
+     *
+     * @return \Generated\Shared\Transfer\CompanyUserTransfer
+     */
+    public function sendInformationMail(CompanyUserTransfer $companyUserTransfer): CompanyUserTransfer;
```
Comment (daniel-rose) [inline bundles/company-user-mail-connector/src/FondOfOryx/Zed/CompanyUserMailConnector/Business/CompanyUserMailConnectorFacadeInterface.php]: Use better naming.

**Review by daniel-rose — APPROVED** [review]

**Review by daniel-rose — APPROVED** [review]

**Pull request merged**